_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,235 @@
+author: Alekseeva Yana
+---
+
+
+## Introduction 
+In [EO](https://github.com/objectionary/eo), caching is used to speed up program compilation.
+Recently we found a caching 
+[bug](https://github.com/objectionary/eo/issues/2790) in `eo-maven-plugin`
+for EO version `0.34.0`. The error occurred because the algorithm compared
+the compilation time and caching time to search for the needed file.
+This is not the most reliable verification method,
+because caching time does not have to be equal to compilation time.
+That we came to the conclusion that we need caching with a reliable verification method
+that does not require reading a file system.
+
+The goal of this blog is to research caching in frequently used build systems (`ccache`, `Maven`, `Gradle`)
+and to create effective caching in [EO](https://github.com/objectionary/eo).
+
+<!--more-->
+
+## Build caching of existing build systems
+
+### ccache/sccache
+In compiled programming languages, building a project takes a long time.
+The reason for the lengthy compilation time is that time is spent on preparing,
+optimizing, checking the code, and so on.
+To speed up the assembly of compiled languages, ccache and sccache are used.
+Let's look at the assembly scheme using C++ as an example
+to imagine the build process in compiled languages:
+
+<p align="center">
+  <img src="/images/ccache.svg">
+</p>
+
+1) First, preprocessor gets the input files. The input files are code files and header files.
+The preprocessor removes comments from the code and converts the code in accordance
+with macros and executes other directives, starting with the “#” symbol
+(such as #include, #define, various directives like #pragma).
+The result is a single edited file with human-readable code that the compiler will get.
+
+
+2) The compiler receives the finished code file and converts it into machine code, presented in an object file.
+At the compilation stage, parsing occurs, which checks whether the code matches
+rules of a specific programming language. Next, the parsing occurs preprocessor code into machine code
+according to the rules.
+At the end of its work, the compiler optimizes the resulting machine code and produces an object file. 
+To speed up compilation, different files of the same project are compiled in parallel,
+that is, we receive several object files at once.
+
+3)  After all received project object files are passed to the linker.
+Linker is a program that combines program components, written in assembly language or a high-level programming language,
+into an executable file or library. The result of the linker is an executable .exe file.
+
+
+As a result, in compiled languages, multiple files are simultaneously and independently converted
+into machine code at the compilation stage.
+This machine code is then combined into one executable file.
+
+
+`ccache` has two main caching methods:
+1) `Direct mode` - hashcode is generated based on the source code.
```

Comment:
@Yanich96 Which "hashcode" do you mean? You gave the definition below, that paragraph positioning confuses a lot. I have to skip this part and then return to it after.